docs: add self signed certs guide

GavinFrazar opened the pull request:
Closes #13781

This PR adds a guide for using self-signed certs with Teleport. It includes a prominent warning about why you should never use self-signed certs in production, and info about how to use `teleport`, `tctl`, and `tsh` with a proxy that uses self-signed certs.

For using `teleport` I included info about passing `--insecure` for helm and for systemd daemon.

The PR also consolidates copies of the same snippet regarding TLS certificate setup into a single include file: 
`tls-certificate-setup.mdx`

GavinFrazar committed:
Add slug for self signed certs

GavinFrazar committed:
Add self-signed certs guide

GavinFrazar committed:
Move tls cert setup into an include snippet

GavinFrazar committed:
Update tsh section on self signed certs

GavinFrazar commented:
I also moved the info on how to setup certs into an include file and included that in the self-signed certs guide

GavinFrazar committed:
Fix indentation

GavinFrazar committed:
Remove extra slash before heading

ptgott commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+If the Teleport Proxy Service is using self-signed certs, then `tsh` will not trust the proxy's certificate.
+In order to use `tsh` in this case, you need to use the `--insecure` flag.
+
+## Configuring Certificates Properly
```
I don't think we need this section, since this guide is explicitly intended for users who have already decided to use self-signed certificates in an insecure environment. If we do want to direct the user to instructions on setting up proper PKI with Teleport, we could, for example, link to https://goteleport.com/docs/deploy-a-cluster/open-source/

GavinFrazar replied:
ha, this is actually a case where I think inlining is warranted since the guide emphasizes the importance of setting up certs properly in prod. I don't feel strongly about it though, so I can replace the inline with a link here

ptgott replied:
I'm not sure this guide is about the importance of setting up certs properly. To me, up until this section, it's explicitly focused on how to use self-signed certificates in non-production environments. If we want to reframe this guide to be about how the Proxy Service uses TLS in general, we could include this section, but we should also:
- move this to a different section of the docs, e.g., `/docs/architecture`
- change the title and introductory section to be about how the Proxy Service uses TLS in general
- make this more of a conceptual guide than a how-to guide

GavinFrazar replied:
This is related to the link discussion we had above, and I think in this case I'll just remove the inline guide and provide a link under "Further Reading" at the bottom of the page.

GavinFrazar committed:
Update docs/config.json

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Fix formatting of github suggestion commit

GavinFrazar committed:
Update docs/pages/management/admin.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar commented on docs/pages/management/admin.mdx:
```diff
@@ -16,6 +16,7 @@ cluster maintenance tasks.
 
 - [Teleport Daemon](./admin/daemon.mdx): Set up Teleport as a daemon on Linux with systemd.
 - [Upgrade the Teleport Binary](./admin/upgrading-the-teleport-binary.mdx): Upgrade the `teleport` binary without losing connections.
+- [Run Teleport with self-signed certs](./admin/self-signed-certs.mdx): Run Teleport with self-signed certs.
```
😅 Yes, it's hard to summarize without getting too verbose, but how is this: 
```diff
-Run Teleport with self-signed certs.
+Set up Teleport in a local environment without configuring TLS certificates.
```

I considered "dev environment" but idk if that works. Basically "an environment for trying out teleport that isn't production". Not sure.

GavinFrazar committed:
Update docs/pages/includes/tls-certificate-setup.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+---
+
+In this guide, we will show you how to run Teleport with self-signed certs, with remedies for problems caused
+by using self-signed certs.
```
I included that in the warning <Notice> below but I could repeat it here?

GavinFrazar replied:
Nvm, I'll move it from the warning up to the intro where it makes more sense.

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+In this guide, we will show you how to run Teleport with self-signed certs, with remedies for problems caused
+by using self-signed certs.
+
+The Teleport Proxy Service authenticates itself to clients via TLS x509 certificates.
```
The proxy service runs the web UI and proxies API calls - in both cases it authenticates to users via x509 certs signed by a CA that the user's machine trust, like from LetsEncrypt for example. So you need the certs configured for secure use of the web UI or `tsh`

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+
+The Teleport Proxy Service authenticates itself to clients via TLS x509 certificates.
+If certificates are not configured for the Proxy Service, then it uses self-signed certificates,
+which clients will not trust by default. This can cause problems when using tools like
```
All of these tools are used to interact with a Teleport cluster.
```suggestion
which clients will not trust by default. This can cause problems interacting with your Teleport cluster via Teleport cli tools, such as
```

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+<Notice type="warning" title="DO NOT USE THIS IN PRODUCTION">
+  **DO NOT USE SELF-SIGNED CERTS IN PRODUCTION**
+
+  It is not recommended to run Teleport in production with self-signed certificates.
```
I'd like to keep the link to the proper configuration section below, but sure let's remove the wiki link to MITM attacks

GavinFrazar replied:
removing link to the config section below which described how to setup TLS certs, since I'm removing that section

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Change link title to avoid abbreviation

GavinFrazar committed:
Update intro to explain the purpose of the guide and why anyone would want it

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+
+## Prerequisites
+
+(!docs/pages/includes/edition-prereqs-tabs.mdx!)
```
Good point, and the Teleport Cloud tab doesnt make sense since in the context of this article either. I'll remove that include and make a custom one for this guide

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+
+    <TabItem label="Running teleport via helm chart">
+    For deploying a cluster, set
+    [extraArgs](../../reference/helm-reference/teleport-cluster.mdx#extraargs)
```
Sure, but for completion let's keep the link and add a nested tab for showing how to do it inline

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+    You also need to run `sudo systemctl daemon-reload` and then `sudo systemctl restart teleport.service`
+    for your change to take effect.
+    
+    See [create a systemd service](./daemon.mdx#step-23-create-and-configure-a-systemd-service) for
```
Why are we avoiding links in guides? Links are informative, concise, keep the guide from veering off-topic, and provide a single source of truth. If I inline an explanation for how to setup a systemd unit file, then that info has to be kept in sync with the systemd guide. It's also more information than I'd like to include in this guide. If I omit the link and don't inline, then the reader who doesn't know how to setup systemd has to go find that guide themselves.

We should be cautious of being overzealous in avoiding links. In this case, I've provided the additional info a user needs to modify their systemd unit file, assuming they know what a unit file is and how systemd services work. If they don't know then I provide a link to our systemd guide. I think that's worth including as a "see more info" link, because anyone who knows what I'm talking about already won't click the link, and anyone who doesn't know can click the link.

Having typed up my argument for links, maybe I should rephrase the sentence around links like this: `If you are unfamiliar with X, you can learn about it here: <link>`

ptgott replied:
The [style guide](https://goteleport.com/docs/contributing/documentation/style-guide/#expectations) advises against adding links to the body sections of how-to guides because the goal of these guides is to get the reader up and running with a particular setup as quickly as possible. 

If we want to include a link in a how-to guide, I think we should make it explicit what a reader should get out of the link, e.g., something concrete like a configuration value, JSON document to copy/paste, etc. If we're telling a reader to consult a link to learn more information, we're not providing them with a quick path toward completing a goal.

GavinFrazar replied:
> If we're telling a reader to consult a link to learn more information, we're not providing them with a quick path toward completing a goal.

I see, in that case I agree we should remove the link here. I can always put "further reading" links at the bottom of the page.

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,117 @@
+
+</Tabs>
+
+## Self-Signed Certs: `tctl`
```
Sounds good

GavinFrazar committed:
Move motivation for guide out of warning and into intro

GavinFrazar committed:
Simplify the warning notice

GavinFrazar committed:
Change headings

GavinFrazar committed:
Capitalize helm and simplify label

GavinFrazar committed:
Update Helm info to include inline instructions

GavinFrazar committed:
Simplify systemd label

GavinFrazar committed:
Update systemd tab

GavinFrazar committed:
Make 'proxy' and 'certs' consistent throughout the guide:
certs -> certificates
proxy -> Proxy Service

GavinFrazar committed:
Update slug certs -> certificates

GavinFrazar committed:
Fix include snippet yaml example

GavinFrazar committed:
Change prereqs to omit cloud tab and instruct user to check for self-signed certs config

GavinFrazar commented:
@ptgott this is ready for re-review, could you take another look?

ptgott commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,216 @@
+When running `tctl` remotely via the Teleport Proxy Service, if the Proxy Service is using self-signed 
+certificates, then `tctl` will not trust the certificate from the Proxy Service.
+To disable certificate verification use:
+`tctl --insecure`.
```
I think we can make "use: `tctl --insecure`" more precise. Right now, it's not clear if this is a full command or not. Maybe say, "use the `--insecure` flag when running `tctl` commands"?

GavinFrazar replied:
Changed according to your suggestion. I also added a "For example: `tctl status --insecure`" at the end of tctl section, and "For example: `tsh login --proxy=tele.example.com:443 --user=alice --insecure`" at the end of the tsh section.

ptgott commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,216 @@
+certificates, then `tctl` will not trust the certificate from the Proxy Service.
+To disable certificate verification use:
+`tctl --insecure`.
+`tctl` will determine how to connect to the auth service in a few ways:
```
```suggestion

`tctl` will determine how to connect to the Auth Service in a few ways:
```
I've suggested capitalizing "Auth Service" and adding a new paragraph since this is the start of a new thought.

ptgott commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,216 @@
+To disable certificate verification use:
+`tctl --insecure`.
+`tctl` will determine how to connect to the auth service in a few ways:
+- loading configuration from the profile after logging in with `tsh`
```
```suggestion
- loading configuration from a local profile after logging in with `tsh`
```
Since we haven't mentioned "profile" yet, I've suggested adding an indefinite article and qualifying this a bit more.

ptgott commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,216 @@
+    </TabItem>
+
+    <TabItem label="systemd">
+    Edit the `systemd` Teleport unit file to include `--insecure` in the `ExecStart` line, for example:
```
Maybe add an instruction to run `systemctl status teleport` to get the path to the unit file?

GavinFrazar replied:
Good idea, added

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Apply suggestions from code review

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar committed:
Update docs/pages/management/admin/self-signed-certs.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,216 @@
+which clients will not trust by default.
+When visiting the cluster's Web UI, the certificate presented will not be trusted by your browser. In this case, 
+you will likely see a page warning you that the website is not trusted.
+Additionally, self-signed certificates can cause problems when interacting with your Teleport cluster via
```
You pretty much hit the nail on the head that the problem in general is going to be connecting to the proxy service, so I'll change this to be more specific.

GavinFrazar committed:
Get specific about problem caused by insecure TLS

GavinFrazar committed:
Remove 'how to setup tls certs' section in favor of a link at the bottom

GavinFrazar committed:
Add examples for tsh and tctl, and clean up tctl wording

GavinFrazar committed:
Remove link to systemd guide

GavinFrazar committed:
Add more detail to systemd instructions

GavinFrazar committed:
Change wording around instructions for teleport

GavinFrazar committed:
Rearrange links

GavinFrazar commented:
@ptgott I incorporated your latest feedback, please take a look when you can. Sorry this is taking up so much of your time

ptgott reviewed (approved):
Thanks!

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/self_signed_certs

GavinFrazar committed:
Remove reference to deleted acme.mdx

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/self_signed_certs

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/self_signed_certs

ryanclark commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,224 @@
+the Teleport Proxy Service endpoint and the Proxy Service is using self-signed certificates, then `teleport` will need
+to be run with the `--insecure` flag to disable verification of the
+Proxy Service TLS certificate. This is the case when:
+- The Teleport config file `auth_servers` setting is set to the Proxy Service endpoint:
+  - `auth_servers: ["tele.example.com:443"]` or
+  - `auth_servers: ["tele.example.com:3080"]`
```
We added config v3 which has either `auth_server` or `proxy_server` 

https://github.com/gravitational/teleport/blob/bd0ae3819b27e107153398de926392c96f8088ae/docs/pages/reference/config.mdx?plain=1#L142-L158

Can you update this to match please?

Note, we didn't backport this, so when backporting for v10 you should still mention `auth_servers`.

GavinFrazar replied:
Sure. On that note, are we planning to update the `teleport start` cli to have separate `--proxy-server` and `--auth-server`?

ryanclark replied:
cc @zmb3

alexfornuto commented:
I disabled auto-merge and reassigned back to @GavinFrazar to address the latest review.

GavinFrazar committed:
Remove v2 config reference to auth_servers

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/self_signed_certs

ryanclark commented:
@GavinFrazar did you mean to assign this PR to me?

GavinFrazar committed:
Remove list syntax for proxy_server

GavinFrazar commented on docs/pages/management/admin/self-signed-certs.mdx:
```diff
@@ -0,0 +1,224 @@
+Proxy Service TLS certificate. This is the case when:
+- The Teleport config file `proxy_server` setting is set to the Proxy Service endpoint:
+  - `proxy_server: ["tele.example.com:443"]` or
+  - `proxy_server: ["tele.example.com:3080"]`
```
thanks for catching this, I wasn't thinking 🥲

GavinFrazar commented:
> @GavinFrazar did you mean to assign this PR to me?

I reassigned after addressing your feedback, since I've seen that a few times in other PRs I thought maybe it was just the norm lol.

GavinFrazar commented:
@ryanclark just waiting on your approval, was there something else you wanted addressed still?

GavinFrazar committed:
Merge branch 'master' into gavinfrazar/self_signed_certs
